homeassistant/components/netgear/device_tracker.py

```diff
@@ -1,166 +1,141 @@
+    router = hass.data[DOMAIN][entry.unique_id]
+    tracked = set()
+
+    @callback
+    def update_router():
+        """Update the values of the router."""
+        add_entities(router, async_add_entities, tracked)
+
+    router.listeners.append(
+        async_dispatcher_connect(hass, router.signal_device_new, update_router)
+    )
+
+    update_router()
+
+
+@callback
+def add_entities(router, async_add_entities, tracked):
+    """Add new tracker entities from the router."""
+    new_tracked = []
+
+    for mac, device in router.devices.items():
+        if mac in tracked:
+            continue
+
+        new_tracked.append(NetgearDevice(router, device))
+        tracked.add(mac)
+
+    if new_tracked:
+        async_add_entities(new_tracked, True)
+
+
+class NetgearDevice(ScannerEntity):
+    """Representation of a Netgear device."""
+
+    def __init__(self, router: NetgearRouter, device: Dict[str, any]) -> None:
+        """Initialize a Netgear device."""
+        self._router = router
+        self._name = device["name"]
+        self._mac = device["mac"]
+        self._manufacturer = device["device_model"]
+        self._icon = icon_for_device(device)
+        self._active = True
+        self._attrs = {}
+
+    def update(self) -> None:
+        """Update the Netgear device."""
+        device = self._router.devices[self._mac]
+        self._active = device["active"]
+        self._attrs = {
+            "link_type": device["type"],
+            "link_rate": device["link_rate"],
+            "signal_strength": device["signal"],
+        }
+        if not self._active:
+            self._attrs = {}
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self._mac
+
+    @property
+    def name(self) -> str:
+        """Return the name."""
+        return self._name
+
+    @property
+    def is_connected(self):
+        """Return true if the device is connected to the network."""
+        return self._active
+
+    @property
+    def source_type(self) -> str:
+        """Return the source type."""
+        return SOURCE_TYPE_ROUTER
+
+    @property
+    def icon(self) -> str:
+        """Return the icon."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self) -> Dict[str, any]:
+        """Return the attributes."""
+        return self._attrs
+
+    @property
+    def device_info(self) -> Dict[str, any]:
+        """Return the device information."""
+        return {
+            "connections": {(CONNECTION_NETWORK_MAC, self._mac)},
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "name": self.name,
+            "manufacturer": self._manufacturer,
+        }
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed."""
+        return False
+
+    async def async_on_demand_update(self):
+        """Update state."""
+        self.async_schedule_update_ha_state(True)
```

Comment:
This method should be a `@callback`, it should contain all the code that is currently in `def update()` and should end with `self.async_write_ha_state()`.

Reply:
I've seen `async def async_on_demand_update(self):` always used like that, since there is only 6 files using it 2 each.
I think I should find a better solution for not using it.

Reply:
Having `homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity Apple TV`

Reply:
Found a solution without using `async_on_demand_update`, clean I think

Reply:
We should not use schedule update ha state if there is no update function to call. If you are getting a `NoEntitySpecifiedError`, it means that you are not attaching the handler in `async_added_to_hass`.